examples/main/main.cpp

```diff
@@ -851,6 +895,84 @@ int main(int argc, char ** argv) {
+                    const std::vector<std::string> cmd = params.special_cmds
+                        ? try_parse_command(buffer)
+                        : std::vector<std::string>();
+
+                    if (cmd.size() == 2 && cmd[0] == "/readfile") {
+                        const std::string filename = cmd[1];
+                        LOG_DBG("reading file: '%s'\n", filename.c_str());
+                        std::ifstream text_file(filename);
+                        if (!text_file) {
+                            LOG("failed to open file '%s'\n", filename.c_str());
+                            continue;
+                        }
+                        pending_input << text_file.rdbuf() << "\n\n";
+                        LOG("read %zu characters from file\n", (size_t) text_file.tellg());
+                        continue;
+                    } else if (cmd.size() == 2 && cmd[0] == "/savesess") {
+                        const std::string filename = cmd[1];
+                        LOG("save session file: '%s'\n", filename.c_str());
+                        size_t res = llama_state_save_file(ctx, filename.c_str(), embd_inp.data(), n_past);
+                        if (res == 0) {
+                            LOG("failed to save session file '%s'\n", filename.c_str());
+                        }
+                        continue;
+                    } else if (cmd.size() == 2 && cmd[0] == "/loadsess") {
+                        const std::string filename = cmd[1];
+                        LOG("load session file: '%s'\n", filename.c_str());
+                        session_tokens.resize(n_ctx);
+                        size_t n_token_count_out;
+                        size_t res = llama_state_load_file(ctx, filename.c_str(), session_tokens.data(), session_tokens.size(), &n_token_count_out);
+                        if (res == 0) {
+                            LOG("failed to load session file '%s'\n", filename.c_str());
+                        } else {
+                            session_tokens.resize(n_token_count_out);
+                            embd_inp = session_tokens;
+                            n_past = n_token_count_out;
+                            llama_kv_cache_seq_rm(ctx, 0, n_past, -1);
+                            LOG("loaded %zu tokens from session file '%s'\n", n_token_count_out, filename.c_str());
+                        }
+                        continue;
+                    } else if (cmd.size() == 1 && cmd[0] == "/regen") {
+                        if (pos_history.empty()) {
+                            LOG("no previous assistant message to regenerate\n");
+                            continue;
+                        }
+                        int last_n_past = pos_history.back();
+                        int n_tokens_removed = n_past - last_n_past;
+                        llama_kv_cache_seq_rm(ctx, 0, last_n_past, -1);
+                        n_remain += n_tokens_removed;
```

Comment:
This might become a problem with `n_predict == -1 (infinite) or -2 (stop at context size)`, comment from [here](https://github.com/ggerganov/llama.cpp/blob/1716e6b25a665cea377281c1439728253761cb4b/examples/main/main.cpp#L1039).